llama_index/indices/struct_store/retriever.py

```diff
@@ -0,0 +1,63 @@
+import logging
+from typing import Any, List, Optional, Union
+
+from sqlalchemy import Table
+
+from llama_index.data_structs.node import NodeWithScore
+from llama_index.indices.base_retriever import BaseRetriever
+from llama_index.indices.query.schema import QueryBundle
+from llama_index.indices.struct_store.sql_table import SQLTableIndex
+
+DEFAULT_QUERY_TMPL = (
+    "Please return the relevant tables (including the full schema) "
+    "for the following query: {orig_query_str}"
+)
+
+logger = logging.getLogger(__name__)
+
+
+class SQLTableRetriever(BaseRetriever):
+    """Retriever for SQL Table Index."""
+
+    def __init__(
+        self,
+        index: SQLTableIndex,
+        query_tmpl: Optional[str] = None,
+        tables: Optional[List[Union[str, Table]]] = None,
+        **kwargs: Any
+    ) -> None:
+        self._index = index
+        self._query_tmpl = query_tmpl or DEFAULT_QUERY_TMPL
+        self._tables = tables
+
+    def _retrieve(
+        self,
+        query_bundle: QueryBundle,
+    ) -> List[NodeWithScore]:
+        """Retrieve nodes."""
+        vector_store_index = self._index.vector_store_index
+
+        # Specific table case. Use initialized tables if specified.
+        if self._tables:
```

Comment:
i'm not actually sure what this is supposed to do. if `self._tables` is not specified, then you do retrieval from the existing index based on the query, which is fair. If it is specified...you feed the schema in and get back a similar document? I'm confused on the use case

Reply:
Hmm yeah so the problem is we lose reference to the individual nodes once they are written into the store, so I need to do a query on the store to get the right schema back.